docs/proposals/31-inplace-node-update.md

```diff
@@ -0,0 +1,388 @@
+
+### Goals
+
+- Provide functionality to update the nodes in-place for Kubernetes minor versions and Machine image versions.
+- Two strategies for in-place - `InPlaceUpdate` and `InplaceUpdateOnLabel`.
+- Maximize re-use of the existing functionality from `machine-controller-manager` and [`gardener-node-agent`](https://github.com/gardener/gardener/blob/master/docs/concepts/node-agent.md).
+
+### Non-Goals
+
+- Provide a way to schedule the pods on the same node after the in-place update.
+- Downgrading of Kubernetes/Machine image version after successful update.
+- Switching between in-place and rolling update strategies.
+- Provide an update strategy without draining the node.
+- Provide in-place updates only for the Kubernetes version in worker pools where the OS does not support in-place updates.
+
+## Proposal
+
+In short, Gardener should support new update strategies that do not require deleting and recreating the node when the Kubernetes minor or Machine image version is updated.
+
+### Approach
+
+The `machine-controller-manager` and `gardener-node-agent` components will be used to perform in-place updates. `machine-controller-manager` will provide an interface that prepares the machine/node for an update. This interface can be used by other components to determine if a node is ready for an update, allowing the other component to perform the update on the node and then notify the `machine-controller-manager` that the node has been successfully updated. _In the case of Gardener, this component will be the `gardener-node-agent`._
+
+This design was chosen to leverage the `machine-controller-manager`'s node rollout orchestration functionality while ensuring that it remains a standalone component.
+
+### Prerequisites
+
+The only prerequisite is that the machine image/OS on the node must support in-place updates and provide a tool or utility to initiate the OS update. This update tool/utility should meet the following requirements:
+
+1. **Failure Recovery**: In case of a failure, the OS should be able to revert to the previous version.
+2. **Retry Configuration**: Support for configuring the number of update retries.
+3. **Registry Configuration**: Ability to specify the registry from which to pull the OS image.
+
+### Update Strategies
+
+Gardener will introduce two additional update strategies, `InPlaceUpdate` and `InPlaceUpdateOnLabel`, for worker pools, in addition to the existing `RollingUpdate`. The two new strategies are described as follows:
```

Comment:
@vlerenc  apologies, but don't see why we need to rename our current strategy. As of today, our `MachineDeploymentStrategyType` of `RollingUpdate` satisfies the k8s rolling update of no downtime (if configured correctly). We are using the same semantics are the k8s `Deployment` whose `Spec.Strategy.Type` is also worded as `RollingUpdate`.  And we have mostly followed the same semantics.  People who see the same word will expect the same semantics.

After all, for a k8S `Deployment`, `Pods` are also terminated and recreated, but they still call it `RollingUpdate` and not `TerminateAndRecreateUpdateRolling` . It makes sense to stick to the same literal word.  

No objections to the revised words for the strategies: `InPlaceUpdateRolling` and `InPlaceUpdateOnLabel`.  Though to be honest, applying the same logic, even the latter can include the word `Rolling`, right ?   Following same logic, it should be `InPlaceRollingUpdate` and `InPlaceRollingUpdateOnLabel`.   At-least it flows off the tongue better that way and is more readable.

Reply:
So much to say:
1.) I like to avoid near-useless renaming effort as well.
2.) Then again, good names, especially if exposed to end-users in the Gardener API, make our lives easier later on. So far, the `MachineDeploymentStrategyType` is only internal (MCM API) - we can keep it/map it. But exposing now strategies with "too wide/ambiguous" names to end users in our Gardener API seems like something we should like to disambiguate/make more clear if possible.
3.) "Deployments" and "MachineDeployments" are not the same thing, e.g. because pods aren't updated in place (even changing the requests in-place was a difficult thing), but machines soon will be (which incidentally is why we are in this situation now with "MachineDeployments" and Kubernetes isn't with "Deployments").
4.) But the main point really is that "MachineDeployment"s and the whole thought of "similarity" to which you spoke is hidden at the Gardener API. Hence, no user will make the connection anyway. As far as they see it, they are asked to define the strategy for how to update workers in a worker pool.
5.) As for "rolling" in combination with "recreate" or "inplace", you cannot know what the user is up to. Maybe she/he just wants to take us out of the picture and there is no real rolling at all - it's more a blue-green update they steer with the deletion of a machine or its label (new nodes are added, but old nodes were tainted and will go out of live over time). We should not assume, users will do the same as we do when we roll (recreate or in-place), when they are in control. We can only roll, because that's all we know how to do, but users may have other ideas what to do with the means we expose to them. 

So, when this GEP writes "Gardener will introduce [] additional update strategies..." I see the Gardener API (this is the gardener repo after all), but you possibly the MCM API? At the Gardener API level I really like to avoid now introducing a strategy name (newly), just because we introduced it in former years at the MCM level (probably needlessly back then, because there was no other strategy anyhow) that nobody ever saw at Gardener API level.

So, at Gardener API level, all strategies are new, so let's give them good names, please. That's no renaming for me - it's exposing a completely new API for worker update strategies in the shoot API.